google-cloud-bigtable/test/test_proxy/test_proxy.rb

```diff
@@ -0,0 +1,449 @@
+        project_id: project_id,
+        timeout: timeout
+      }.compact
+
+      if data_target == "emulator"
+        args[:emulator_host] = ENV.fetch("BIGTABLE_EMULATOR_HOST")
+      else
+        args[:endpoint] = data_target
+        args[:credentials] = GRPC::Core::Channel.new data_target, nil, :this_channel_is_insecure
+      end
+
+      Google::Cloud::Bigtable.new(**args)
+    end
+  end
+
+  # Creates a client in the proxy.
+  # Each client has its own dedicated channel(s), and can be used concurrently
+  # and independently with other clients.
+  def create_client req, _call
+    LOGGER.info "CreateClient(#{req.to_h})"
+    Client.create req
+    tp::CreateClientResponse.new
+  end
+
+  # Closes a client in the proxy, making it not accept new requests.
+  def close_client req, _call
+    LOGGER.info "CloseClient(#{req.to_h})"
+    Client.get(req.client_id).close
+    tp::CloseClientResponse.new
+  end
+
+  # Removes a client in the proxy, making it inaccessible. Client closing
+  # should be done by CloseClient() separately.
+  def remove_client req, _call
+    LOGGER.info "RemoveClient(#{req.to_h})"
+    Client.remove req.client_id
+    tp::RemoveClientResponse.new
+  end
+
+  # Reads a row with the client instance.
+  # The result row may not be present in the response.
+  # Callers should check for it (e.g. calling has_row() in C++).
+  # rpc ReadRow(ReadRowRequest) returns (RowResult) {}
+  def read_row req, _call
+    LOGGER.info "ReadRow(#{req.to_h})"
+
+    result = Client.get(req.client_id)
+                   .table(req.table_name)
+                   .read_row(req.row_key)
+
+    if result.nil?
+      # TODO(meagar): Is this the right response?
+      #   It seems to allow the tests to pass...
+      tp::RowResult.new(
+        status: not_found_status,
+        row: bt::V2::Row.new(key: req.row_key)
+      )
+    else
+      tp::RowResult.new(
+        # The status is not available from the Ruby client library
+        status: ok_status,
+        row: row_to_v2_row(result)
+      )
+    end
+  rescue Google::Cloud::Error => e
+    LOGGER.info "ReadRow failed: Caught #{e}"
+    abort "Unexpected codepath reached, populate the error handler for ReadRow"
+  end
+
+  # Reads rows with the client instance.
+  # rpc ReadRow(ReadRowRequest) returns (RowResult) {}
+  # rubocop:disable Metrics/AbcSize
+  def read_rows req, _call
+    LOGGER.info "ReadRows(#{req.to_h})"
+
+    # At the time of writing, no tests set this field; rather than guess at an implementation,
+    # this is a note for future maintainers to implement this if the field becomes set in the test suite
+    assert req.request.app_profile_id == "", "An app profile ID was specified when non was expected"
+
+    # TODO: req.request.reverse is not supported by the client library; when it is, fix this
+    result = Client.get(req.client_id)
+                   .table(req.request.table_name)
+                   .read_rows(keys: req.request&.rows&.row_keys&.to_a)
```

Comment:
looks like rows limit is not passed in? And also row ranges? Maybe this is why the TestReadRows_NoRetry_ErrorAfterLastRow failed?